File: tools/assemble_plugin_jar.py
#!/usr/bin/env python3
"""Assemble the brikk-sql compiler plugin into one jar for `-Xplugin=` consumption.

Merges (first wins on duplicate entries):
  build/tasks/_brikk-sql-compiler-plugin_jarJvm/brikk-sql-compiler-plugin-jvm.jar
  build/tasks/_brikk-sql_jarJvm/brikk-sql-jvm.jar
  kotlinx-serialization-core-jvm-<ver>.jar   (located in the Kotlin Toolchain's dependency cache)

into build/brikk-sql-compiler-plugin-<kotlin>-<lib>.jar  (default: ...-2.4.0-0.1.0.jar).

The file name deliberately follows KEFS's default detect+version pattern
`<artifact-id>-<kotlin-version>-<lib-version>.jar` so the IDE plugin recognises the
`-Xplugin=` reference with no replacement patterns (docs/vendor/kefs/GUIDE.md §6).

Run after `./kotlin build -m brikk-sql-compiler-plugin`. This is a plain merge, not a
relocating shade: fine for the in-repo smoke module, NOT what a published (KEFS-loadable)
plugin needs — see docs/vendor/kefs/PLUGIN_AUTHORS.md ("dependencies must be relocated").
"""
from __future__ import annotations

import glob
import os
import sys
import zipfile

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
ARTIFACT = "brikk-sql-compiler-plugin"
KOTLIN_VERSION = "2.4.0"   # compiler the plugin is compiled against
LIB_VERSION = "0.1.0"
VERSION = f"{KOTLIN_VERSION}-{LIB_VERSION}"
OUT = os.path.join(ROOT, "build", f"{ARTIFACT}-{VERSION}.jar")

LOCAL_JARS = [
    os.path.join(ROOT, "build", "tasks", "_brikk-sql-compiler-plugin_jarJvm", "brikk-sql-compiler-plugin-jvm.jar"),
    os.path.join(ROOT, "build", "tasks", "_brikk-sql_jarJvm", "brikk-sql-jvm.jar"),
]

CACHE_ROOTS = [
    os.environ.get("KOTLIN_TOOLCHAIN_M2_CACHE", ""),
    os.path.expanduser("~/.cache/JetBrains/Kotlin/.m2.cache"),
    os.path.expanduser("~/.m2/repository"),
]

DEP_GLOBS = [
    "org/jetbrains/kotlinx/kotlinx-serialization-core-jvm/*/kotlinx-serialization-core-jvm-*[0-9].jar",
]


def find_dep(pattern: str) -> str:
    for root in CACHE_ROOTS:
        if not root:
            continue
        hits = sorted(glob.glob(os.path.join(root, pattern)))
        hits = [h for h in hits if not h.endswith(("-sources.jar", "-javadoc.jar"))]
        if hits:
            return hits[-1]
    sys.exit(f"dependency not found in any cache root: {pattern}")


def main() -> None:
    jars = list(LOCAL_JARS)
    for j in jars:
        if not os.path.exists(j):
            sys.exit(f"missing {j}\nrun: ./kotlin build -m brikk-sql-compiler-plugin")
    jars += [find_dep(p) for p in DEP_GLOBS]

    os.makedirs(os.path.dirname(OUT), exist_ok=True)
    seen: set[str] = set()
    with zipfile.ZipFile(OUT, "w", zipfile.ZIP_DEFLATED) as out:
        for jar in jars:
            with zipfile.ZipFile(jar) as src:
                for info in src.infolist():
                    name = info.filename
                    if name in seen or name.endswith("/"):
                        continue
                    if name.startswith("META-INF/") and (name.endswith((".SF", ".RSA", ".DSA")) or name == "META-INF/MANIFEST.MF"):
                        continue
                    seen.add(name)
                    out.writestr(info, src.read(name))
    print(f"wrote {os.path.relpath(OUT, ROOT)} from {len(jars)} jars ({len(seen)} entries)")


if __name__ == "__main__":
    main()

File: brikk-sql-plugin-tooling/src/dev.brikk.house.sql.tooling/AssemblePluginJar.kt
package dev.brikk.house.sql.tooling

import java.nio.file.Files
import java.nio.file.Path
import java.util.zip.ZipEntry
import java.util.zip.ZipFile
import java.util.zip.ZipOutputStream
import kotlin.io.path.createDirectories
import kotlin.io.path.name
import kotlin.io.path.outputStream
import kotlin.io.path.relativeTo
import org.jetbrains.amper.plugins.Classpath
import org.jetbrains.amper.plugins.CompilationArtifact
import org.jetbrains.amper.plugins.Input
import org.jetbrains.amper.plugins.Output
import org.jetbrains.amper.plugins.TaskAction

/**
 * Merges the compiler-plugin jar with its runtime dependencies (brikk-sql, brikk-sql-metadata,
 * kotlinx-serialization) into one jar for `-Xplugin=` consumption.
 *
 * The output is named `<artifactId>-<kotlinVersion>-<libVersion>.jar`: KEFS detects a compiler
 * plugin from the `-Xplugin` jar *file name*, matched as `<detect>-<version>.jar` with the
 * default `<kotlin-version>-<lib-version>` version pattern, so this name needs no remapping in
 * the KEFS settings.
 *
 * This is a plain merge (first entry wins), not a relocating shade. Fine for the in-repo smoke
 * module and the KEFS experiment; NOT what a published plugin needs — see
 * docs/vendor/kefs/PLUGIN_AUTHORS.md ("dependencies must be relocated").
 */
@TaskAction
fun assemblePluginJar(
    @Input pluginJar: CompilationArtifact,
    @Input runtimeClasspath: Classpath,
    artifactId: String,
    kotlinVersion: String,
    libVersion: String,
    bundleExcludes: List<String>,
    @Output outputDir: Path,
) {
    val dependencyJars = runtimeClasspath.resolvedFiles
        .filter { it != pluginJar.artifact && it.name.endsWith(".jar") }
        .filterNot { jar -> bundleExcludes.any { jar.name.startsWith(it) } }
        .sortedBy { it.name }
    val sources = listOf(pluginJar.artifact) + dependencyJars

    val out = outputDir.resolve("$artifactId-$kotlinVersion-$libVersion.jar")
    outputDir.toFile().deleteRecursively()
    outputDir.createDirectories()

    val seen = HashSet<String>()
    ZipOutputStream(out.outputStream().buffered()).use { zip ->
        for (source in sources) {
            ZipFile(source.toFile()).use { jar ->
                for (entry in jar.entries().asSequence()) {
                    val name = entry.name
                    if (entry.isDirectory || name in seen || isSigningOrManifest(name)) continue
                    seen += name
                    zip.putNextEntry(ZipEntry(name).also { it.time = entry.time })
                    jar.getInputStream(entry).use { it.copyTo(zip) }
                    zip.closeEntry()
                }
            }
        }
    }
    println("wrote ${out.relativeTo(outputDir.parent.parent)} from ${sources.size} jars (${seen.size} entries)")
    println("bundled: ${sources.joinToString { it.name }}")
}

/** Signature files and the top-level manifest never survive a merge. Nested service files do. */
private fun isSigningOrManifest(name: String): Boolean =
    name == "META-INF/MANIFEST.MF" ||
        (name.startsWith("META-INF/") && name.substringAfterLast('.', "").uppercase() in setOf("SF", "RSA", "DSA", "EC"))

internal fun singleJarIn(dir: Path): Path {
    require(Files.isDirectory(dir)) { "missing $dir - run `./kotlin do assemblePluginJar` first" }
    val jars = Files.list(dir).use { s -> s.filter { it.name.endsWith(".jar") }.toList() }
    return jars.singleOrNull() ?: error("expected exactly one jar in $dir, found ${jars.map { it.name }}")
}
